app/build.gradle

```diff
@@ -466,3 +468,34 @@ if (project.hasProperty("coverage")) {
 afterEvaluate {
     check.dependsOn 'findbugs', 'pmd', 'checkstyle'
 }
+
+// -------------------------------------------------------------------------------------------------
+// Bitbar Cloud Testing: testDroidUpload
+// Upload both the application and test APK and conduct a UI test-run
+// Utilizes a preset group of Amazon Fire TV devices on a preset project and framework
+// -------------------------------------------------------------------------------------------------
+
+testdroid {
+
+    try {
+        def tokenFile = new File("${rootDir}/.bitbar_token.json")
+        def parsedFile = new groovy.json.JsonSlurper().parseText(tokenFile.text)
+
+        apiKey parsedFile.secret.bitbarToken.api_key
+        cloudUrl parsedFile.secret.bitbarToken.cloud_url
+
+        println "(Added from .bitbar_token file)"
+    } catch (FileNotFoundException ignored) {
+        println("X_X")
```

Comment:
Is it possible for this task to succeed if the file is not found?  I feel like we should probably throw an exception and cancel the build if not.  

If it is possible for this to succeed, we should flesh out this log so that someone who is debugging this problem can easily trace the log back to this task.  Maybe something like "BitBar credentials not found, canceling remote UI test".

Reply:
This file is necessary (as it contains the API key and cloud URL), otherwise a build won't be uploaded. I changed `ignored` to `e` and clarified the message. What else is needed?

Reply:
I think failing to load it should be considered a build failure, so in the `catch` block I would add `throw(e)`.  I assume that a task throwing an exception should cancel the rest of the build?

Reply:
Should be yeah; added.

Reply:
It's possible that I'm being overly nitpicky here, but I think the ideal situation would be to have this task only run on PRs.  When we're building for a PR, it seems like this running successfully **_should_** be a hard requirement (please correct me if I'm wrong there, you understand the process much better than I do).  But I agree that blocking local builds on this would be unnecessary.

What do you think?

Reply:
It's up to the team really. Originally I had it on PRs, I think that's where it makes sense to be. @npark-mozilla thoughts?

Reply:
I probably am not understanding the full conversation thread, but just to clarify, the bitbar testing will only run on merge to master, and not on PR.  And, in case the file is not found, it will fail almost instantly on the next line when it tries to log in, with invalid credential.  I don't think throw() would be too useful here, and for simplicity, I think having a detailed println statement as above should be sufficient.

Reply:
Agreed. So to clarify once more

- Currently, it runs on merge to master (via task cluster.yml)
- If for whatever reason credentials can't be fetched, we'll see the printed line at least
- A green build will still happen (which is fine for now).
- Results can be checked in Slack. In the case of failure, I suppose you'll notice no results.

Does this sound good?

Reply:
So only thing I'm kinda worried about is that it returns green build after failure.  Devs probably won't give a 2nd look to check for the test result if it is marked green.

 If it was a TC shell script, then you can make it return nonzero int exit value (`exit 1`)  to signal failure on TC.  But because we're running it as a gradle command, it'll be difficult to do that.  
If we can somehow write a file that shows whether the UI test run is successful or not, then run a shell script as a part of TC run that checks this file to return zero/nonzero int, then we can fail the TC run appropriately.  If that is difficult, we can leave it as is, and create a followup issue where we either use their rest api or something else where we can at least know whether there was a failure.

Reply:
Yeah, a green build on failure is definitely a big problem.

Reply:
The only form of direct callback AFAIK is the call back URL *after* a test run has finished. Bitbar testing sends a POST message to a defined URL when test run is finished.

E.g, `testRunId=%s&projectId=%s&status=FINISHED`

The URL has to be reachable online.

We can continue to explore that if we want?

http://docs.bitbar.com/testing/cloud-integrations/jenkins-plugin/jenkins-ui/index.html?highlight=hook

Reply:
Actually, also if the testdroid returns any pass/fail status, we can try to fail the gradle run by raising gradleexception as well. https://stackoverflow.com/questions/10312259/recommended-way-to-stop-a-gradle-build

Reply:
Or, much easier way would be to call the rest api method with the testrunid we got from above to get the test result.
 We can do that on a separate script and return nonzero if it contains failure.

Reply:
After investigation on Sunday it looks like `Status` (from the POST parameter value: http://ptsv2.com/t/xuzfr-1550774903/d/30001)  will only provide us details about a test run’s overall state; in the case of failure it still finishes, hence `FINISHED`. Regardless of failures it will always value as `FINISHED` (or maybe `Waiting` if devices are offline). 

If we want test failure count, we need the test `runID`. We have a set `projectID` we can re-use. 

We can still use the hook callback URL but ignore that state and look at the other parameters (http://ptsv2.com/t/xuzfr-1550774903/d/30001). We would need some kind of service to wait for, poll, and read POST data after a test-run to a server online. From there we get their generated `runID `. With the `runID` and `projectID` that is also available, we can get JSON data back (e.g, Python requests) that will have `"failedTestCaseCount”` and any non zero integer s a failure. From there we can do whatever. But suddenly you introduce a lot of wait (~15 min) and a lot of complication.

Reply:
I put a question to bitbar team on the slack channel about getting the test result - I think it is sufficient to get whether tests have all passed or not.  Anything more detailed than that they can check in on bitbar.  Currently the problem is that bitbar just says `Finished` even when there is a failure.  I have an idea that we can try later.  IF that wait is the wait during the actual test execution, I think that is okay.  Other TC test runs don't finish until tests are completed anyways.  If this doesn't work, we can ask Levente to check the slack channel for failure, since slack notification tells whether there are failures or not.